File: rust/kansei-core/src/geometries/geometry.rs
use bytemuck::{Pod, Zeroable};

/// A single vertex attribute descriptor.
#[derive(Debug, Clone)]
pub struct VertexAttribute {
    pub shader_location: u32,
    pub offset: u64,
    pub format: wgpu::VertexFormat,
}

/// Standard interleaved vertex: position(vec4) + normal(vec3) + uv(vec2).
#[repr(C)]
#[derive(Debug, Clone, Copy, Pod, Zeroable)]
pub struct Vertex {
    pub position: [f32; 4],
    pub normal: [f32; 3],
    pub uv: [f32; 2],
}

impl Vertex {
    pub const LAYOUT: wgpu::VertexBufferLayout<'static> = wgpu::VertexBufferLayout {
        array_stride: std::mem::size_of::<Vertex>() as u64,
        step_mode: wgpu::VertexStepMode::Vertex,
        attributes: &[
            // @location(0) position: vec4<f32>
            wgpu::VertexAttribute {
                format: wgpu::VertexFormat::Float32x4,
                offset: 0,
                shader_location: 0,
            },
            // @location(1) normal: vec3<f32>
            wgpu::VertexAttribute {
                format: wgpu::VertexFormat::Float32x3,
                offset: 16,
                shader_location: 1,
            },
            // @location(2) uv: vec2<f32>
            wgpu::VertexAttribute {
                format: wgpu::VertexFormat::Float32x2,
                offset: 28,
                shader_location: 2,
            },
        ],
    };
}

/// A GPU geometry — vertex + index buffers.
pub struct Geometry {
    pub label: String,
    pub vertices: Vec<Vertex>,
    pub indices: Vec<u32>,
    pub vertex_buffer: Option<wgpu::Buffer>,
    pub index_buffer: Option<wgpu::Buffer>,
    /// Optional indirect draw args buffer (DrawIndexedIndirect: 5 × u32).
    pub indirect_args_buffer: Option<wgpu::Buffer>,
    pub initialized: bool,
    // External buffer pointers (for compute-generated geometry, e.g. marching cubes)
    ext_vertex_buffer: Option<*const wgpu::Buffer>,
    ext_index_buffer: Option<*const wgpu::Buffer>,
    ext_indirect_buffer: Option<*const wgpu::Buffer>,
}

// SAFETY: wgpu::Buffer is internally Arc-based and Send+Sync.
// The raw pointers are only dereferenced on the same thread that created them.
unsafe impl Send for Geometry {}
unsafe impl Sync for Geometry {}

impl Geometry {
    /// Get the active vertex buffer (owned or external).
    pub fn active_vertex_buffer(&self) -> Option<&wgpu::Buffer> {
        self.vertex_buffer.as_ref().or_else(|| {
            self.ext_vertex_buffer.map(|p| unsafe { &*p })
        })
    }
    /// Get the active index buffer (owned or external).
    pub fn active_index_buffer(&self) -> Option<&wgpu::Buffer> {
        self.index_buffer.as_ref().or_else(|| {
            self.ext_index_buffer.map(|p| unsafe { &*p })
        })
    }
    /// Get the active indirect args buffer (owned or external).
    pub fn active_indirect_buffer(&self) -> Option<&wgpu::Buffer> {
        self.indirect_args_buffer.as_ref().or_else(|| {
            self.ext_indirect_buffer.map(|p| unsafe { &*p })
        })
    }
}

impl Geometry {
    pub fn new(label: &str, vertices: Vec<Vertex>, indices: Vec<u32>) -> Self {
        Self {
            label: label.to_string(),
            vertices,
            indices,
            vertex_buffer: None,
            index_buffer: None,
            indirect_args_buffer: None,
            initialized: false,
            ext_vertex_buffer: None,
            ext_index_buffer: None,
            ext_indirect_buffer: None,
        }
    }

    /// Create a Geometry placeholder for externally-owned GPU buffers (zero readback).
    /// Initially has no buffer pointers — call `set_external_buffers()` once the
    /// owning struct is at its final heap address.
    pub fn new_indirect_placeholder(label: &str) -> Self {
        Self {
            label: label.to_string(),
            vertices: Vec::new(),
            indices: Vec::new(),
            vertex_buffer: None,
            index_buffer: None,
            indirect_args_buffer: None,
            initialized: true,
            ext_vertex_buffer: None,
            ext_index_buffer: None,
            ext_indirect_buffer: None,
        }
    }

    pub fn initialize(&mut self, device: &wgpu::Device) {
        use wgpu::util::DeviceExt;

        self.vertex_buffer = Some(device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
            label: Some(&format!("{}/Vertices", self.label)),
            contents: bytemuck::cast_slice(&self.vertices),
            usage: wgpu::BufferUsages::VERTEX,
        }));

        self.index_buffer = Some(device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
            label: Some(&format!("{}/Indices", self.label)),
            contents: bytemuck::cast_slice(&self.indices),
            usage: wgpu::BufferUsages::INDEX,
        }));

        self.initialized = true;
    }

    pub fn index_count(&self) -> u32 {
        self.indices.len() as u32
    }

    pub fn vertex_count(&self) -> u32 {
        self.vertices.len() as u32
    }

    /// Update external buffer pointers. Call after the owning struct is at its final
    /// heap address (e.g., after moving into Rc<RefCell<>>).
    ///
    /// # Safety
    /// Pointers must remain valid for the lifetime of this Geometry.
    pub unsafe fn set_external_buffers(
        &mut self,
        vertex: *const wgpu::Buffer,
        index: *const wgpu::Buffer,
        indirect: Option<*const wgpu::Buffer>,
    ) {
        self.ext_vertex_buffer = Some(vertex);
        self.ext_index_buffer = Some(index);
        self.ext_indirect_buffer = indirect;
    }

    /// Returns true if this geometry uses GPU-driven indirect drawing.
    pub fn is_indirect(&self) -> bool {
        self.active_indirect_buffer().is_some()
    }
}

File: src/buffers/Geometry.ts
/**
 * Represents a geometric mesh with vertex and index buffers for WebGPU rendering.
 */
class Geometry {
    /** Indicates if this geometry is used for instanced rendering */
    public isInstancedGeometry: boolean = false;

    /** WebGPU buffer containing vertex data */
    public vertexBuffer?: GPUBuffer;

    /** WebGPU buffer containing index data */
    public indexBuffer?: GPUBuffer;

    /** Format of the index buffer data */
    public indexFormat: GPUIndexFormat = "uint16";

    /** Collection of vertex buffer layout descriptors */
    public vertexBuffersDescriptors: Iterable<GPUVertexBufferLayout | null> = [];

    /** Indicates if the geometry has been initialized with GPU buffers */
    public initialized: boolean = false;

    /** Number of vertices in the geometry */
    public vertexCount: number = 0;

    /** Raw vertex data containing interleaved positions, normals, and UVs */
    public vertices?: Float32Array;

    /** Raw index data for defining triangles */
    public indices?: Uint16Array | Uint32Array;

    /** Optional indirect draw args buffer (DrawIndexedIndirect: 5 × u32).
     *  When set, the renderer uses drawIndexedIndirect() instead of drawIndexed(). */
    public indirectArgsBuffer?: GPUBuffer;

    /** True if vertex/index buffers are externally owned (not created by initialize()). */
    public isExternal: boolean = false;

    constructor() { }

    /**
     * Create a Geometry that wraps externally-owned GPU buffers (zero readback).
     * Used for compute-generated meshes (e.g. marching cubes) where the vertex/index data
     * lives entirely on the GPU. The `indirectArgsBuffer` holds DrawIndexedIndirect args
     * so the triangle count is also GPU-driven.
     *
     * The vertex layout matches the standard engine Vertex format:
     *   position(vec4) + normal(vec3) + uv(vec2) — 36-byte stride.
     */
    public static fromGpuBuffers(
        vertexBuffer: GPUBuffer,
        indexBuffer: GPUBuffer,
        indirectArgsBuffer: GPUBuffer,
        indexFormat: GPUIndexFormat = 'uint32',
    ): Geometry {
        const geo = new Geometry();
        geo.vertexBuffer = vertexBuffer;
        geo.indexBuffer = indexBuffer;
        geo.indirectArgsBuffer = indirectArgsBuffer;
        geo.indexFormat = indexFormat;
        geo.isExternal = true;
        geo.initialized = true;
        (geo.vertexBuffersDescriptors as Array<GPUVertexBufferLayout>).push({
            attributes: [
                { shaderLocation: 0, offset: 0, format: 'float32x4' },
                { shaderLocation: 1, offset: 16, format: 'float32x3' },
                { shaderLocation: 2, offset: 28, format: 'float32x2' },
            ],
            arrayStride: 36,
            stepMode: 'vertex',
        });
        return geo;
    }

    /**
     * Initializes the geometry by creating GPU buffers and setting up vertex layouts.
     * @param gpuDevice - The WebGPU device to create buffers on
     */
    public initialize(gpuDevice: GPUDevice) {
        if (this.isExternal) { this.initialized = true; return; }
        this.vertexBuffer = gpuDevice.createBuffer({
            label: 'vertex buffer',
            size: this.vertices!.byteLength,
            usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST,
            mappedAtCreation: true
        });
        new Float32Array(this.vertexBuffer.getMappedRange()).set(this.vertices!);
        this.vertexBuffer.unmap();

        this.indexBuffer = gpuDevice.createBuffer({
            label: 'index buffer',
            size: Math.ceil(this.indices!.byteLength / 4) * 4,
            usage: GPUBufferUsage.INDEX | GPUBufferUsage.COPY_DST,
            mappedAtCreation: true
        });
        const IndexArray = this.indexFormat === "uint32" ? Uint32Array : Uint16Array;
        new IndexArray(this.indexBuffer.getMappedRange()).set(this.indices!);
        this.indexBuffer.unmap();

        (this.vertexBuffersDescriptors as Array<GPUVertexBufferLayout>).push({
            attributes: [
                {
                    shaderLocation: 0 as GPUIndex32,
                    offset: 0 as GPUSize64,
                    format: "float32x4" as GPUVertexFormat
                },
                {
                    shaderLocation: 1 as GPUIndex32,
                    offset: 4 * 4 as GPUSize64,
                    format: "float32x3" as GPUVertexFormat
                },
                {
                    shaderLocation: 2 as GPUIndex32,
                    offset: 4 * 4 + 4 * 3 as GPUSize64,
                    format: "float32x2" as GPUVertexFormat
                }
            ] as Iterable<GPUVertexAttribute>,
            arrayStride: 4 * 4 + 4 * 3 + 4 * 2 as GPUSize32,
            stepMode: "vertex" as GPUVertexStepMode
        });

        this.initialized = true;
    }
}

export { Geometry }
